File: dashboard/src/lib/socketManager.ts
import { io, Socket } from 'socket.io-client';
import { SOCKET_URL, getToken } from './api';

const EVENTS = [
  'device:online',
  'device:offline',
  'device:location',
  'device:battery',
  'device:status',
] as const;

type EventHandler = (event: string, data: unknown) => void;

interface SocketAck {
  ok: boolean;
  error?: string;
  [key: string]: unknown;
}

let socket: Socket | null = null;
let connectPromise: Promise<Socket> | null = null;

function createSocket(token?: string | null): Socket {
  return io(SOCKET_URL, {
    auth: token ? { token } : {},
    transports: ['websocket', 'polling'],
    reconnection: true,
    reconnectionAttempts: Infinity,
    reconnectionDelay: 1000,
    timeout: 15000,
    autoConnect: false,
  });
}

function connect(): Socket | null {
  const token = getToken();

  if (socket) {
    const currentToken = (socket.auth as { token?: string })?.token;
    if (currentToken !== token) {
      socket.disconnect();
      socket = null;
      connectPromise = null;
    }
  }

  if (!socket) {
    socket = createSocket(token);
  }

  return socket;
}

function connectGuest(): Socket {
  if (socket?.connected) {
    return socket;
  }

  if (socket) {
    socket.disconnect();
  }

  socket = createSocket(null);
  connectPromise = null;
  return socket;
}

function waitForConnection(s: Socket): Promise<Socket> {
  if (s.connected) return Promise.resolve(s);

  if (!connectPromise) {
    connectPromise = new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        reject(new Error('Socket connection timeout'));
      }, 15000);

      const onConnect = () => {
        clearTimeout(timeout);
        s.off('connect_error', onError);
        resolve(s);
      };

      const onError = (err: Error) => {
        clearTimeout(timeout);
        s.off('connect', onConnect);
        reject(err);
      };

      s.once('connect', onConnect);
      s.once('connect_error', onError);
      if (!s.connected) s.connect();
    });
  }

  return connectPromise;
}

export async function socketRequest<T extends SocketAck>(
  event: string,
  data?: Record<string, unknown>,
  options?: { guest?: boolean }
): Promise<T> {
  const s = options?.guest ? connectGuest() : connect();
  if (!s) throw new Error('Not authenticated');

  const connected = await waitForConnection(s);

  return new Promise((resolve, reject) => {
    connected.timeout(15000).emit(event, data ?? {}, (response: T) => {
      if (response?.ok) {
        resolve(response);
      } else {
        reject(new Error(response?.error || 'Socket request failed'));
      }
    });
  });
}

export function disconnectSocket() {
  if (socket) {
    socket.disconnect();
    socket = null;
    connectPromise = null;
  }
}

export function subscribe(handler: EventHandler): () => void {
  const s = connect();
  if (!s) return () => {};

  void waitForConnection(s).catch(() => {});

  const listeners = EVENTS.map((event) => {
    const fn = (data: unknown) => handler(event, data);
    s.on(event, fn);
    return { event, fn };
  });

  return () => {
    listeners.forEach(({ event, fn }) => s.off(event, fn));
  };
}

File: android-app/app/src/main/java/com/devicemonitor/app/data/socket/SocketManager.kt
package com.devicemonitor.app.data.socket

import com.devicemonitor.app.BuildConfig
import com.devicemonitor.app.data.prefs.TokenManager
import io.socket.client.Ack
import io.socket.client.IO
import io.socket.client.Socket
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withTimeoutOrNull
import org.json.JSONObject
import kotlin.coroutines.resume

class SocketManager(private val tokenManager: TokenManager) {

    private var socket: Socket? = null
    private var onReconnectListener: (() -> Unit)? = null

    fun setOnReconnectListener(listener: (() -> Unit)?) {
        onReconnectListener = listener
    }

    @Synchronized
    fun connect(): Socket {
        val token = tokenManager.getToken()
            ?: throw IllegalStateException("Not authenticated")

        if (socket?.connected() == true) {
            return socket!!
        }

        if (socket != null && !socket!!.connected()) {
            socket!!.connect()
            return socket!!
        }

        disconnect()

        val options = IO.Options().apply {
            auth = mapOf("token" to token)
            reconnection = true
            reconnectionAttempts = Int.MAX_VALUE
            reconnectionDelay = 1000
            reconnectionDelayMax = 5000
            transports = arrayOf("websocket", "polling")
            timeout = 20000
        }

        socket = IO.socket(BuildConfig.SOCKET_URL, options).apply {
            on(Socket.EVENT_CONNECT) {
                onReconnectListener?.invoke()
            }
        }
        socket!!.connect()
        return socket!!
    }

    @Synchronized
    fun connectGuest(): Socket {
        disconnect()

        val options = IO.Options().apply {
            reconnection = true
            transports = arrayOf("websocket", "polling")
            timeout = 20000
        }

        socket = IO.socket(BuildConfig.SOCKET_URL, options)
        socket!!.connect()
        return socket!!
    }

    @Synchronized
    fun disconnect() {
        socket?.off(Socket.EVENT_CONNECT)
        socket?.disconnect()
        socket?.off()
        socket = null
    }

    fun isConnected(): Boolean = socket?.connected() == true

    @Synchronized
    fun ensureConnected(): Socket {
        return if (isConnected()) socket!! else connect()
    }

    /**
     * Waits up to [CONNECT_WAIT_MS] for the socket to become connected before emitting.
     * This prevents the "socket timeout" that happens when emit fires before the
     * TCP/WebSocket handshake completes.
     */
    private suspend fun waitForConnection(sock: Socket): Boolean {
        if (sock.connected()) return true
        return withTimeoutOrNull(CONNECT_WAIT_MS) {
            suspendCancellableCoroutine { cont ->
                sock.once(Socket.EVENT_CONNECT) { cont.resume(true) }
                sock.once(Socket.EVENT_CONNECT_ERROR) { cont.resume(false) }
            }
        } ?: false
    }

    suspend fun emitAck(event: String, data: JSONObject, guest: Boolean = false): JSONObject {
        return withTimeoutOrNull(EMIT_TIMEOUT_MS) {
            suspendCancellableCoroutine { cont ->
                try {
                    val activeSocket = when {
                        guest -> connectGuest()
                        else -> ensureConnected()
                    }

                    // If socket is not yet connected, wait for connection before emitting.
                    // Without this, emit fires immediately and the ack never arrives because
                    // the underlying transport isn't ready yet — causing a false timeout.
                    if (!activeSocket.connected()) {
                        activeSocket.once(Socket.EVENT_CONNECT) {
                            activeSocket.emit(event, data, Ack { args ->
                                val response = args.firstOrNull() as? JSONObject
                                    ?: JSONObject().put("ok", false).put("error", "Empty response")
                                if (cont.isActive) cont.resume(response)
                            })
                        }
                        activeSocket.once(Socket.EVENT_CONNECT_ERROR) { args ->
                            val err = args.firstOrNull()?.toString() ?: "Connection failed"
                            if (cont.isActive) cont.resume(
                                JSONObject().put("ok", false).put("error", err)
                            )
                        }
                    } else {
                        activeSocket.emit(event, data, Ack { args ->
                            val response = args.firstOrNull() as? JSONObject
                                ?: JSONObject().put("ok", false).put("error", "Empty response")
                            if (cont.isActive) cont.resume(response)
                        })
                    }
                } catch (e: Exception) {
                    cont.resume(
                        JSONObject()
                            .put("ok", false)
                            .put("error", e.message ?: "Socket error")
                    )
                }
            }
        } ?: JSONObject().put("ok", false).put("error", "Socket timeout")
    }

    companion object {
        private const val EMIT_TIMEOUT_MS = 20_000L   // increased from 8s — gives time to connect + get ack
        private const val CONNECT_WAIT_MS = 15_000L   // max wait for socket handshake before emitting
    }
}
